rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,306 @@
+            case None           => StateT.liftF(Stream.Empty)
+            case Some(filtered) => listMatch(filtered, prem, matcher, merger, varLevel, wildcard)
+          }
+        } else {
+          for {
+            trem        <- listMatchItem(targets, pattern, matcher)
+            forcedYield <- listMatch(trem, prem, matcher, merger, varLevel, wildcard)
+          } yield forcedYield
+        }
+      }
+    }
+
+  def listMatchItem[T](tlist: Seq[T],
+                       pattern: T,
+                       matcher: (T, T) => OptionalFreeMap[Unit]): NonDetFreeMap[Seq[T]] =
+    for {
+      triple               <- StateT.liftF(singleOut(tlist).toStream)
+      (head, target, tail) = triple
+      forcedYield <- StateT[Stream, FreeMap, Seq[T]]((s: FreeMap) => {
+                      matcher(target, pattern).run(s) match {
+                        case None =>
+                          Stream.Empty
+                        case Some((state, _)) =>
+                          Stream((state, head.reverse ++ tail))
+                      }
+                    })
+    } yield forcedYield
+
+  def spatialMatch(target: Par, pattern: Par): OptionalFreeMap[Unit] =
+    if (pattern.freeCount === 0) {
+      if (pattern == target)
+        StateT.pure(Unit)
+      else {
+        StateT.liftF(None)
+      }
+    } else {
+      val varLevel: Option[Int] = possiblyFind[Expr, Int](
+        { (expr) =>
+          expr.exprInstance match {
+            case EVarBody(EVar(Some(v))) =>
+              v.varInstance match {
+                case FreeVar(level) => Some(level)
+                case _              => None
+              }
+            case _ => None
+          }
+        },
+        pattern.exprs
+      )
+      val wildcard: Boolean = exists[Expr](
+        { (expr) =>
+          expr.exprInstance match {
+            case EVarBody(EVar(Some(v))) =>
+              v.varInstance match {
+                case Wildcard(_) => true
+                case _           => false
+              }
+            case _ => false
+          }
+        },
+        pattern.exprs
+      )
+
+      def NoFrees(exprs: Seq[Expr]): Seq[Expr] =
+        exprs.filter({ (expr) =>
+          expr.exprInstance match {
+            case EVarBody(EVar(Some((v)))) =>
+              v.varInstance match {
+                case FreeVar(_)  => false
+                case Wildcard(_) => false
+                case _           => true
+              }
+            case _ => true
+          }
+        })
+
+      for {
+        _ <- listMatchSingle[Send](target.sends,
+                                   pattern.sends,
+                                   spatialMatch,
+                                   (p, s) => p.withSends(s +: p.sends),
+                                   varLevel,
+                                   wildcard)
+        _ <- listMatchSingle[Expr](target.exprs,
+                                   NoFrees(pattern.exprs),
```

Comment:
So if I understand this correctly, we remove the frees and then we try to match "the rest". And if the rest match, we fold/assign whatever was not "the rest" to the first capture variable or wildcard (with preference to the capture variable as you wrote in the comment).

So for something like
```
target = 5 | 6 | 10 | 11

pattern = 5 | x | 10 | y
```
we would get
```
x = 6 | 11
```
instead of
```
x = 6
y = 11
```

Reply:
Yes. I don't see why that's a problem. matching with more than one variable in parallel is non-deterministic. We just take the first one.

Reply:
So I think 

5 | 6 | 10 | 11

matching with

5 | 6 | x | y

where x = 10 | 11 and y = none is fine. But I think its confusing when it is across sorted boundaries.

Reply:
Resolved through chat. Basically for this matcher implementation we can just give a warning saying

x | y is non-determinate